pipeline-model-definition/src/main/groovy/org/jenkinsci/plugins/pipeline/modeldefinition/Utils.groovy

```diff
@@ -320,12 +321,45 @@ class Utils {
                     tagsAction.addTag(tagName, tagValue)
                     currentNode.save()
                 }
+            }
+        }
+    }
+
+    static void markStartAndEndNodesInStageAsNotExecuted(String stageName, FlowExecution execution = null) {
+        if (execution == null) {
+            CpsThread thread = CpsThread.current()
+            execution = thread.execution
+        }
+
+        List<BlockStartNode> nodes = []
+
+        ForkScanner scanner = new ForkScanner()
+
+        FlowNode stage = scanner.findFirstMatch(execution.currentHeads, null, isStageWithOptionalName(stageName))
 
-                if (isRestart) {
-                    currentNode.addAction(new NotExecutedNodeAction())
+        if (stage != null && stage instanceof BlockStartNode) {
+            nodes.add(stage)
+
+            // Additional check needed to get the possible enclosing parallel branch for a nested stage.
+            Filterator<FlowNode> filtered = FlowScanningUtils.fetchEnclosingBlocks(stage)
```

Comment:
So my big uncertainty here is if we *do* need to limit to just the enclosing parallel branch or can do a more global search for matching branch names.  I *think* so but I haven't traced out all the legal combinations possible as of sequential stages (like we should check we can't get false positives from similarly named branches/stages that can occur elsewhere).

Reply:
That is a possibility - stages are guaranteed to have unique names *in their context* but are not guaranteed to have unique names across other nested stage (parallel or sequential) contexts.